lib/Echidna/Campaign.hs

```diff
@@ -241,20 +244,91 @@ addToCorpus :: (MonadState s m, Has Campaign s) => Integer -> [(Tx, (VMResult, I
 addToCorpus n res = unless (null rtxs) $ hasLens . corpus %= DS.insert (n, rtxs) where
   rtxs = map fst res
 
-seqMutators :: (MonadRandom m) => MutationConsts -> m (Int -> Corpus -> [Tx] -> m [Tx])
-seqMutators (c1, c2, c3) = fromList 
-                            [(cnm      , fromInteger c1),
-                             (mut False, fromInteger c2),
-                             (mut True , fromInteger c3)]
-  where -- Use the generated random transactions
-        cnm _ _          = return
-        mut flp ql ctxs gtxs = do
-          let somePercent = if (fst . DS.findMax) ctxs > 1    -- if the corpus already contains new elements 
-                             then 1 + (DS.size ctxs `div` 20) -- then take 5% of its size
-                             else DS.size ctxs                -- otherwise, take all of it
-          rtxs <- fromList $ map (\(i, txs) -> (txs, fromInteger i)) $ take somePercent $ DS.toDescList ctxs
-          k <- getRandomR (0, length rtxs - 1)
-          return . take ql $ if flp then take k gtxs ++ rtxs else take k rtxs ++ gtxs
+data TxsMutation = TxIdentity
+                 | Shrinking
+                 | Mutation
+                 | Expansion
+                 | Swapping
+                 | Deletion
+
+data CorpusMutation = Skip
+                    | RandomAppend TxsMutation 
+                    | RandomPrepend TxsMutation
+                    | RandomSplice
+                    | RandomInterleave
+
+selectAndMutate :: MonadRandom m 
+                => ([Tx] -> m [Tx]) -> Corpus -> m [Tx]
+selectAndMutate f ctxs = do
```

Comment:
is there a point for passing this function as argument if it's a tail call anyway?

Reply:
There is no functional reason for this, just to avoid the function application to be repeated after each use of `selectAndMutate` (this code was inspired by a `hlint` suggestion on repeated code that I needed to refactor).